File: clients/ios/RemoteEverything/Core/Setup/SetupURI.swift
import Foundation

/// Strict parser for the `remote-everything://setup` invitation URI, mirroring
/// contracts/schemas/setup-uri.schema.json: exactly the known parameters, each
/// exactly once, each shaped as the schema says. An invitation that does not
/// parse cleanly is not an invitation.
///
/// Percent-decoding is hand-rolled on purpose. `URLComponents` decodes while it
/// parses and silently merges repeated parameters, which is the wrong behaviour
/// for a security boundary: a duplicate `node` must be refused, not last-one-wins.
enum SetupURI {

    /// One invitation, as a client reads it.
    struct Invitation: Equatable {
        var node: String
        var nodeName: String
        var origin: String
        var invitation: String
        var serverPin: ServerPin?
    }

    /// Why an invitation was refused. The reason is for logs and tests; the UI
    /// says one thing for every malformed invitation.
    enum Rejected: Error, Equatable {
        case notASetupURI
        case malformedParameter
        case unknownParameter(String)
        case duplicateParameter(String)
        case missingParameter(String)
        case badNodeID
        case badNodeName
        case badOrigin
        case badInvitation
        case badPins
        case badEncoding
    }

    private static let knownParameters: Set<String> = [
        "node", "node_name", "origin", "invitation", "fingerprint", "public_key_pin",
    ]

    private static let requiredParameters = ["node", "node_name", "origin", "invitation"]

    static func parse(_ text: String) throws -> Invitation {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let prefix = "remote-everything://setup"
        guard trimmed.hasPrefix(prefix) else { throw Rejected.notASetupURI }
        let rest = String(trimmed.dropFirst(prefix.count))
        // The host is `setup` and nothing else: a path, a second slash or a
        // fragment is a shape this URI does not have.
        guard rest.hasPrefix("?") else { throw Rejected.notASetupURI }
        let query = String(rest.dropFirst())

        var parameters: [String: String] = [:]
        for pair in query.split(separator: "&", omittingEmptySubsequences: false) {
            let pairText = String(pair)
            guard let separator = pairText.firstIndex(of: "=") else { throw Rejected.malformedParameter }
            let key = String(pairText[pairText.startIndex..<separator])
            guard !key.isEmpty else { throw Rejected.malformedParameter }
            guard knownParameters.contains(key) else { throw Rejected.unknownParameter(key) }
            guard parameters[key] == nil else { throw Rejected.duplicateParameter(key) }
            let rawValue = String(pairText[pairText.index(after: separator)...])
            parameters[key] = try percentDecode(rawValue)
        }

        for required in requiredParameters where parameters[required] == nil {
            throw Rejected.missingParameter(required)
        }

        guard let node = parameters["node"], isHex64(node) else { throw Rejected.badNodeID }
        guard let nodeName = parameters["node_name"], isValidNodeName(nodeName) else { throw Rejected.badNodeName }
        guard let origin = parameters["origin"], isOrigin(origin) else { throw Rejected.badOrigin }
        guard let invitation = parameters["invitation"], isInvitationToken(invitation) else {
            throw Rejected.badInvitation
        }

        // Both halves of a self-signed gateway's pin travel together or not at
        // all: a client told to expect a certificate it cannot check is worse
        // off than one told to trust the system.
        let fingerprint = parameters["fingerprint"]
        let publicKeyPin = parameters["public_key_pin"]
        guard (fingerprint == nil) == (publicKeyPin == nil) else { throw Rejected.badPins }
        var pin: ServerPin?
        if let fingerprint, let publicKeyPin {
            guard isHex64(fingerprint), isBase64SHA256(publicKeyPin) else { throw Rejected.badPins }
            pin = ServerPin(certFingerprint: fingerprint, publicKeyPin: publicKeyPin)
        }

        return Invitation(
            node: node,
            nodeName: nodeName,
            origin: origin,
            invitation: invitation,
            serverPin: pin
        )
    }

    // MARK: - Shapes (setup-uri.schema.json)

    static func isHex64(_ value: String) -> Bool {
        guard value.count == 64 else { return false }
        return value.allSatisfy { character in
            (character >= "0" && character <= "9") || (character >= "a" && character <= "f")
        }
    }

    /// 1–80 Unicode code points, no C0 or DEL control characters, and not
    /// padded with whitespace. Code points, not grapheme clusters: Go counts
    /// runes and the three clients have to agree on the count.
    static func isValidNodeName(_ value: String) -> Bool {
        guard !value.isEmpty, value.unicodeScalars.count <= 80 else { return false }
        guard value == value.trimmingCharacters(in: .whitespacesAndNewlines) else { return false }
        return value.unicodeScalars.allSatisfy { scalar in
            scalar.value >= 0x20 && scalar.value != 0x7F
        }
    }

    /// `https://host[:port]` — scheme, host, optional port; no path, query or
    /// fragment, and no credentials in the authority.
    static func isOrigin(_ value: String) -> Bool {
        let prefix = "https://"
        guard value.hasPrefix(prefix) else { return false }
        let authority = String(value.dropFirst(prefix.count))
        guard !authority.isEmpty else { return false }
        guard !authority.contains("/"), !authority.contains("?"), !authority.contains("#") else { return false }
        guard !authority.contains("@") else { return false }
        let host: Substring
        let port: Substring?
        if let colon = authority.lastIndex(of: ":") {
            host = authority[authority.startIndex..<colon]
            port = authority[authority.index(after: colon)...]
        } else {
            host = authority[...]
            port = nil
        }
        guard !host.isEmpty else { return false }
        if let port {
            guard !port.isEmpty, port.allSatisfy({ $0.isNumber }) else { return false }
            guard let number = Int(port), (1...65535).contains(number) else { return false }
        }
        return true
    }

    /// Base64url of 32 bytes: exactly 43 characters of the URL-safe alphabet.
    static func isInvitationToken(_ value: String) -> Bool {
        guard value.count == 43 else { return false }
        return value.allSatisfy { character in
            (character >= "A" && character <= "Z")
                || (character >= "a" && character <= "z")
                || (character >= "0" && character <= "9")
                || character == "-"
                || character == "_"
        }
    }

    /// Standard base64 of a 32-byte digest: 43 characters plus one `=`.
    static func isBase64SHA256(_ value: String) -> Bool {
        guard value.count == 44, value.hasSuffix("=") else { return false }
        return value.dropLast().allSatisfy { character in
            (character >= "A" && character <= "Z")
                || (character >= "a" && character <= "z")
                || (character >= "0" && character <= "9")
                || character == "+"
                || character == "/"
        }
    }

    // MARK: - Percent decoding

    /// Form decoding: `%XX` is a byte, `+` is a space, everything else passes
    /// through. Bytes are collected first and decoded as UTF-8 once, so a
    /// multi-byte escape survives.
    static func percentDecode(_ value: String) throws -> String {
        let input = Array(value.utf8)
        var output: [UInt8] = []
        output.reserveCapacity(input.count)
        var index = 0
        while index < input.count {
            let byte = input[index]
            if byte == 0x25 { // %
                guard index + 2 < input.count,
                      let high = hexValue(input[index + 1]),
                      let low = hexValue(input[index + 2])
                else { throw Rejected.badEncoding }
                output.append((high << 4) | low)
                index += 3
            } else if byte == 0x2B { // +
                output.append(0x20)
                index += 1
            } else {
                output.append(byte)
                index += 1
            }
        }
        guard let decoded = String(bytes: output, encoding: .utf8) else { throw Rejected.badEncoding }
        return decoded
    }

    private static func hexValue(_ byte: UInt8) -> UInt8? {
        switch byte {
        case 0x30...0x39: return byte - 0x30 // 0-9
        case 0x41...0x46: return byte - 0x41 + 10 // A-F
        case 0x61...0x66: return byte - 0x61 + 10 // a-f
        default: return nil
        }
    }
}

File: clients/android/app/src/main/java/com/remoteeverything/core/setup/SetupUri.kt
package com.remoteeverything.core.setup

import java.net.URI

/**
 * Strict parser for the remote-everything://setup invitation URI, mirroring
 * contracts/setup-uri.schema.json: exactly the known parameters, each exactly
 * once, each shaped as the schema says. An invitation that does not parse
 * cleanly is not an invitation.
 */
object SetupUri {

    data class Invitation(
        val node: String,
        val nodeName: String,
        val origin: String,
        val invitation: String,
        val serverPin: ServerPin?,
    ) {
        data class ServerPin(val certFingerprint: String, val publicKeyPin: String)
    }

    class Rejected(val reason: String) : Exception(reason)

    private val nodeId = Regex("^[a-f0-9]{64}$")
    private val invitationToken = Regex("^[A-Za-z0-9_-]{43}$")
    private val origin = Regex("^https://[^/?#]+$")
    private val hexFingerprint = Regex("^[a-f0-9]{64}$")
    private val spkiPin = Regex("^[A-Za-z0-9+/]{43}=$")
    private val c0OrDel = Regex("[\\u0000-\\u001F\\u007F]")

    private val knownKeys = setOf("node", "node_name", "origin", "invitation", "fingerprint", "public_key_pin")

    fun parse(text: String): Invitation {
        val trimmed = text.trim()
        val uri = try {
            URI(trimmed)
        } catch (e: Exception) {
            throw Rejected("not a URI")
        }
        if (uri.scheme != "remote-everything" || uri.host != "setup") {
            throw Rejected("not a setup URI")
        }
        if (!uri.path.isNullOrEmpty() && uri.path != "/") {
            throw Rejected("setup URI carries a path")
        }
        if (uri.rawFragment != null || trimmed.contains('#')) {
            throw Rejected("setup URI carries a fragment")
        }
        val params = linkedMapOf<String, String>()
        val query = uri.rawQuery ?: throw Rejected("no parameters")
        for (pair in query.split('&')) {
            val eq = pair.indexOf('=')
            if (eq <= 0) throw Rejected("malformed parameter")
            val key = pair.substring(0, eq)
            if (key !in knownKeys) throw Rejected("unknown parameter: $key")
            if (params.put(key, decode(pair.substring(eq + 1))) != null) {
                throw Rejected("duplicate parameter: $key")
            }
        }
        for (required in listOf("node", "node_name", "origin", "invitation")) {
            if (required !in params) throw Rejected("missing parameter: $required")
        }
        val node = params.getValue("node")
        if (!nodeId.matches(node)) throw Rejected("bad node id")
        val nodeName = params.getValue("node_name")
        if (nodeName.isEmpty() || nodeName.codePointCount(0, nodeName.length) > 80 ||
            nodeName != nodeName.trim() || c0OrDel.containsMatchIn(nodeName)
        ) {
            throw Rejected("bad node name")
        }
        val originValue = params.getValue("origin")
        if (!isValidOrigin(originValue)) throw Rejected("bad origin")
        val invitation = params.getValue("invitation")
        if (!invitationToken.matches(invitation)) throw Rejected("bad invitation")
        val fingerprint = params["fingerprint"]
        val pin = params["public_key_pin"]
        if ((fingerprint == null) != (pin == null)) {
            throw Rejected("fingerprint and public_key_pin travel together")
        }
        if (fingerprint != null && !hexFingerprint.matches(fingerprint)) throw Rejected("bad fingerprint")
        if (pin != null && !spkiPin.matches(pin)) throw Rejected("bad public_key_pin")
        return Invitation(
            node = node,
            nodeName = nodeName,
            origin = originValue,
            invitation = invitation,
            serverPin = if (fingerprint != null && pin != null) {
                Invitation.ServerPin(fingerprint, pin)
            } else {
                null
            },
        )
    }

    /**
     * Percent-decoding with form semantics ('+' is a space), hand-rolled on
     * purpose: this is a security boundary, and the platform decoder's charset
     * overload needs API 33. Bytes are collected and decoded as UTF-8 once, so
     * multi-byte escapes work.
     */
    private fun decode(value: String): String {
        val bytes = java.io.ByteArrayOutputStream(value.length)
        fun flushLiteral(from: Int, to: Int) {
            if (to > from) bytes.write(value.substring(from, to).toByteArray(Charsets.UTF_8))
        }
        var i = 0
        var literalStart = 0
        while (i < value.length) {
            when (value[i]) {
                '%' -> {
                    if (i + 2 >= value.length) throw Rejected("bad encoding")
                    flushLiteral(literalStart, i)
                    val byte = value.substring(i + 1, i + 3).toIntOrNull(16) ?: throw Rejected("bad encoding")
                    bytes.write(byte)
                    i += 3
                    literalStart = i
                }
                '+' -> {
                    flushLiteral(literalStart, i)
                    bytes.write(' '.code)
                    i += 1
                    literalStart = i
                }
                else -> i += 1
            }
        }
        flushLiteral(literalStart, value.length)
        return String(bytes.toByteArray(), Charsets.UTF_8)
    }

    private fun isValidOrigin(value: String): Boolean {
        if (!value.startsWith("https://")) return false
        val authority = value.removePrefix("https://")
        if (authority.isEmpty()) return false
        if (authority.contains('/') || authority.contains('?') || authority.contains('#')) return false
        if (authority.contains('@')) return false
        val colon = authority.lastIndexOf(':')
        val host: String
        val portStr: String?
        if (colon >= 0) {
            host = authority.substring(0, colon)
            portStr = authority.substring(colon + 1)
        } else {
            host = authority
            portStr = null
        }
        if (host.isEmpty()) return false
        if (portStr != null) {
            if (portStr.isEmpty() || !portStr.all { it.isDigit() }) return false
            val port = portStr.toIntOrNull() ?: return false
            if (port !in 1..65535) return false
        }
        return true
    }
}
